### PR TITLE
#981: Fixes issue #981 and refactors for modularity and clarity

### LINKED ISSUE #981: Problem combining export sets and module refinement
This example produces an error and it seems it should not.
```
module A {
  export reveals a
  const a := 10;
}

module B refines A {
  export reveals *  // complains duplicate B
}
```

The internal problem seems to be that export set A when copied into module B retains its 'default ness' and but is given the name B, after a check whether an export set B is already present, so there ends up being two B's, and an error. SO an implicit invariant is being violated.

1) In any case, it should certainly be possible to override any inherited default export set
but also
2) in what way and with what name should default export set A be inherited, perhaps
a) as export B extends A
b) as export B reveals * extends A
c) as export B reveals *
?
3) What if the refinement parent has an export set (or any other declaration) that matches the name of the refining module? Keeping the export set namespace separate from other declarations might help but not solve this problem.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #981

### CODE REVIEW

File: Test/git-issues/git-issue-981.dfy [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+module BB refines AA {
+  export BB extends AA // error: No explicit export set AA
+}
+
+// Testing some exports and extension errors
+module A {
+  export reveals a
+  export C reveals a
+  const a := 10;
+  const D := 20;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981.dfy:56]:
```suggestion
  const a := 10
  const D := 20
```

File: Test/git-issues/git-issue-981.dfy [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+}
+
+module D refines A {
+  // OK. D in A is not an export set
+}
+
+module E {
+  export E reveals *
+  const F := 40;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981.dfy:89]:
```suggestion
  const F := 40
```

File: Test/git-issues/git-issue-981.dfy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  method F() {} // error: reuse of name F
+  const E := 30; // OK: export sets are in a different namespace
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981.dfy:91]:
```suggestion
  const E := 30 // OK: export sets are in a different namespace
```

File: Test/git-issues/git-issue-981d.dfy [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  export E1 reveals *
+  export E2 reveals *
+  export E3 reveals *
+  export E4 reveals *
+  export E5 reveals *
+  export E6 reveals *
+  module E1 {}
+  import E2 = Z
+  type T3
+  const E4 := 0;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981d.dfy:17]:
```suggestion
  const E4 := 0
```

File: Test/git-issues/git-issue-981d.dfy [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+module B1 {
+  module E1 {}
+  module E3 {}
+  module E4 {}
+  module E5 {}
+  module E6 {}
+  type T3
+  const E4 := 0; // error
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981d.dfy:29]:
```suggestion
  const E4 := 0 // error
```

File: Test/git-issues/git-issue-981d.dfy [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  class E6 {} // error
+}
+
+module C {
+  type T1
+  type T2
+  type T3
+  type T4 = int
+  type T1 // error
+  const T2 := 0; // error
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981d.dfy:40]:
```suggestion
  const T2 := 0 // error
```

File: Test/git-issues/git-issue-981d.dfy [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+module D {
+  const T2 := 0;
+  const T3 := 0;
+  class T4 {}
+  const T2 := 0; // error
+  method T3() {} // error
+  method T3() {} // error
+  const T4 := 0; // error
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981d.dfy:52]:
```suggestion
  const T2 := 0
  const T3 := 0
  class T4 {}
  const T2 := 0 // error
  method T3() {} // error
  method T3() {} // error
  const T4 := 0 // error
```

File: Test/git-issues/git-issue-981e.dfy [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  export E1 reveals *
+  export E2 reveals *
+  export E3 reveals *
+  export E4 reveals *
+  export E5 reveals *
+  export E6 reveals *
+  module E1 {}
+  import E2 = Z
+  type T3
+  const E4 := 0;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981e.dfy:15]:
```suggestion
  const E4 := 0
```

File: Test/git-issues/git-issue-981.dfy [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  module KK {
+    module KKK {
+      const k := 10;
+    }
+  }
+}
+
+// Checks that refines does resolve a qualified ID
+// and does not use a local module
+module L refines K.KK.KKK {
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-981.dfy:14]: At some point, we discussed not allowed qualified names after `refines`. Is that restriction coming in a further PR or did you figure out a good way to deal with these?

File: Source/Dafny/Resolver.cs [hunk 10]
```diff
@@ -1416,56 +1568,76 @@ public class ModuleBindings
...
       dependencies.AddVertex(moduleDecl);
       if (moduleDecl is LiteralModuleDecl) {
         ProcessDependenciesDefinition(moduleDecl, ((LiteralModuleDecl)moduleDecl).ModuleDef, bindings, dependencies);
       } else if (moduleDecl is AliasModuleDecl) {
         var alias = moduleDecl as AliasModuleDecl;
         ModuleDecl root;
-        if (!bindings.TryLookupFilter(alias.Path[0], out root,
+        // Find a match for the root of the alias target that (a) is not the alias itself
+        // and (b) either (b1) is a LiteralModuleDecl or (b2) has the same parent as the alias
+        // I think the intent is to omit non-local aliases -- but why?  TODO
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:1624]: I don't know

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Source/Dafny/Cloner.cs [hunk 11]
```diff
@@ -111,16 +103,17 @@ class Cloner
         }
       } else if (d is ModuleDecl) {
         if (d is LiteralModuleDecl) {
+          // TODO: Does not clone any details; is still resolved
           return new LiteralModuleDecl(((LiteralModuleDecl)d).ModuleDef, m);
         } else if (d is AliasModuleDecl) {
           var a = (AliasModuleDecl)d;
-          return new AliasModuleDecl(a.Path, a.tok, m, a.Opened, a.Exports);
+          return new AliasModuleDecl(a.TargetQId ?? a.TargetQId.clone(false), a.tok, m, a.Opened, a.Exports);
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Cloner.cs:110]: OK

File: Source/Dafny/DafnyAst.cs [hunk 12]
```diff
@@ -3697,12 +3706,83 @@ public class ModuleSignature {
...
+    public string rootName() {
+      return Path[0].val;
+    }
+
+    public IToken rootToken() {
+      return Path[0];
+    }
+
+    override
+    public string ToString() {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/DafnyAst.cs:3741]: OK

File: Source/Dafny/Cloner.cs [hunk 13]
```diff
@@ -111,16 +103,17 @@ class Cloner
         }
       } else if (d is ModuleDecl) {
         if (d is LiteralModuleDecl) {
+          // TODO: Does not clone any details; is still resolved
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Cloner.cs:106]: Done

File: Source/Dafny/RefinementTransformer.cs [hunk 14]
```diff
@@ -76,32 +77,26 @@ public RefinementTransformer(Program p)
...
 
     internal override void PreResolve(ModuleDefinition m) {
-      if (m.RefinementBaseRoot != null) {
-        RefinedSig = m.RefinementBaseRoot.Signature;
+      if (m.RefinementQId?.Decl != null) { // There is a refinement parent and it resolved OK
+        RefinedSig = m.RefinementQId.Sig;
 
         if (RefinedSig.ModuleDef != null) {
-          m.RefinementBase = RefinedSig.ModuleDef;
+          m.RefinementQId.Def = RefinedSig.ModuleDef; // TODO - already set if RefinementQId.Sig is?
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/RefinementTransformer.cs:84]: Done

File: Test/git-issues/git-issue-864rr.dfy [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// Tests the resolution of the refinement parent -- that it ignores
+// local declarations.
+module A {
+  const a := 10;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864rr.dfy:11]: OK

File: Test/git-issues/git-issue-864rr.dfy [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+module B refines A { // ignore the submodule A, use the top-level A
+  module A {
+    const a := 30;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864rr.dfy:16]: OK

File: Test/git-issues/git-issue-864y.dfy [hunk 17]
```diff
@@ -0,0 +1,20 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module B {
+  export B reveals *
+  module A { const a := 30; }
+  const A := 60;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864y.dfy:7]: OK

File: Test/git-issues/git-issue-864z.dfy [hunk 18]
```diff
@@ -0,0 +1,32 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module A {
+  export reveals a
+  const a := 10;
+  const b := 20;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864z.dfy:7]: OK

File: Test/git-issues/git-issue-864z.dfy [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+module B refines A {
+  export reveals *
+  const c := 30;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864z.dfy:12]: OK

File: Test/git-issues/git-issue-864zz.dfy [hunk 20]
```diff
@@ -0,0 +1,12 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module A {
+  export reveals a
+  const a := 10;
+  const b := 20;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-864zz.dfy:7]: OK

File: Test/git-issues/git-issue-981.dfy [hunk 21]
```diff
@@ -0,0 +1,93 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module K {
+  module KK {
+    module KKK {
+      const k := 10;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:7]: OK

File: Test/git-issues/git-issue-981.dfy (hunk 1 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:56]: OK

File: Test/git-issues/git-issue-981.dfy (hunk 3 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:91]: OK

File: Test/git-issues/git-issue-981d.dfy (hunk 4 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981d.dfy:17]: OK

File: Test/git-issues/git-issue-981d.dfy (hunk 5 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981d.dfy:29]: OK

File: Test/git-issues/git-issue-981d.dfy (hunk 7 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981d.dfy:52]: OK

File: Test/git-issues/git-issue-981.dfy (hunk 9 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:14]: I implemented this. I could not come up with a scenario that causes trouble for qualified names vs only sibling modules. The dependencies work out fine.

File: Test/git-issues/git-issue-981e.dfy (hunk 8 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981e.dfy:15]: OK

File: Test/git-issues/git-issue-981d.dfy (hunk 6 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981d.dfy:40]: OK

File: Test/git-issues/git-issue-981.dfy (hunk 2 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:89]: OK

File: Source/Dafny/Resolver.cs (hunk 10 above)
Comment (davidcok) [reply Source/Dafny/Resolver.cs:1624]: Reworked this -- in fact non-local aliases are not used in the resolution of the id root.

File: Test/git-issues/git-issue-981.dfy (hunk 9 above)
Comment (davidcok) [reply Test/git-issues/git-issue-981.dfy:14]: Documented the potential problem in #994, for the next round of work.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**